mautic/mautic
# Add symfony messenger support

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | tba
| Related developer documentation PR URL | tba
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 




#### Description:
This is integration of https://symfony.com/doc/4.4/messenger.html
Add new option to send emails by Mautic 

![image](https://user-images.githubusercontent.com/462477/130192460-9a556957-864d-461b-8914-499986c48d10.png)

### How to use?

At the moment there is a issue with composer https://github.com/mautic/mautic/pull/10293
That means you need call `composer require symfony/messenger`  in your console.





#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Run `composer require symfony/messenger`  because this issue https://github.com/mautic/mautic/pull/10293
3. Go to Configuration > Emails > How should email be handled?   set to Messenger Transport
4. Set your Messenger Transport DSN (for test ou can create RabbitMQ for free https://www.cloudamqp.com/)
5. Send email
6. Email should not be send, should wait in  queue
7. Run `php bin/console messenger:consume email`
8. Email should be send now
9. If you need add more options you can configure it by  `messenger_transport_email_options` in local.php. For example you can can change queue (by default it's `messages`

```
    'messenger_transport_email_options' => [
        'queues' => [
            'emails' => [],
        ],
    ],
```
If you change queue then your command could be  `php bin/console messenger:consume email --queues=emails`
10. Also you can change retry strategy If send email throw expection.  You can set it by   `messenger_transport_email_retry_strategy` in local.php, by default it's (https://symfony.com/doc/4.4/messenger.html#retries-failures)

```
'messenger_transport_email_retry_strategy' => [
                    'max_retries' => 3,
                    // milliseconds delay
                    'delay' => 1000,
                    // causes the delay to be higher before each retry
                    // e.g. 1 second delay, 2 seconds, 4 seconds
                    'multiplier' => 2,
                    'max_delay' => 0,
                    // override all of this with a service that
                    // implements Symfony\Component\Messenger\Retry\RetryStrategyInterface
                    // 'service' => null,
                ],
        ],
```

After these 3 failures, contact is marked as DNC (standard behavior also for immediately and queue mode)

For more details you can follow https://symfony.com/doc/4.4/messenger.html




<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10255"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
